File: python/list-ops/list_ops.py
def append(xs, ys):
    xs_length = length(xs)
    ys_length = length(ys)

    result = [None] * (xs_length + ys_length)
    result[:xs_length] = xs
    result[xs_length:] = ys

    return result


def concat(lists):
    result = []
    for one_list in lists:
        result = append(result, one_list)
    return result


def filter_clone(function, xs):
    return [x for x in xs if function(x)]


def length(xs):
    result = 0
    for _ in xs:
        result += 1
    return result


def map_clone(function, xs):
    return [function(x) for x in xs]


def foldl(function, xs, acc):
    result = acc
    for x in xs:
        result = function(result, x)
    return result


def foldr(function, xs, acc):
    result = acc
    for x in reverse(xs):
        result = function(x, result)
    return result


def reverse(xs):
    xs_length = length(xs)

    result = [None] * xs_length
    i = 0
    while i != xs_length:
        result[i] = xs[xs_length - 1 - i]
        i += 1
    return result

File: javascript/list-ops/list-ops.js
//
// This is only a SKELETON file for the 'List - Ops' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class List {
  constructor(a, b) {
    if (a === undefined) {
      this.last = null;
    } else if (a instanceof List) {
      this.prefix = a;
      this.last = b;
    } else {
      const length = b === undefined ? a.length : b;

      if (length === 0) {
        return new List();
      } else {
        return new List(new List(a, length - 1), a[length - 1]);
      }
    }
  }

  get values() {
    if (this.isEmpty()) {
      return [];
    }

    const result = this.prefix.values;
    result.push(this.last);

    return result;
  }

  isEmpty() {
    return this.last === null;
  }

  push(value) {
    return new List(this, value);
  }

  append(list) {
    return list.isEmpty() ? this : this.append(list.prefix).push(list.last);
  }

  concat(lists) {
    return lists.isEmpty()
      ? this
      : this.concat(lists.prefix).append(lists.last);
  }

  filter(callback) {
    return this.isEmpty()
      ? this
      : callback(this.last)
      ? new List(this.prefix.filter(callback), this.last)
      : this.prefix.filter(callback);
  }

  map(callback) {
    return this.isEmpty()
      ? this
      : new List(this.prefix.map(callback), callback(this.last));
  }

  length() {
    return this.isEmpty() ? 0 : this.prefix.length() + 1;
  }

  foldl(callback, initialValue) {
    return this.reverse().foldr(callback, initialValue);
  }

  foldr(callback, initialValue) {
    return this.isEmpty()
      ? initialValue
      : this.prefix.foldr(callback, callback(initialValue, this.last));
  }

  reverse() {
    return this.isEmpty()
      ? this
      : new List().push(this.last).append(this.prefix.reverse());
  }
}
